linkup-cli/src/services/caddy.rs

```diff
@@ -1,167 +1,190 @@
 use std::{
     fs,
+    path::PathBuf,
     process::{Command, Stdio},
 };
 
 use crate::{
-    linkup_dir_path, linkup_file_path, CliError, Result, LINKUP_CF_TLS_API_ENV_VAR,
-    LINKUP_LOCALSERVER_PORT,
+    linkup_dir_path, linkup_file_path, local_config::LocalState, LINKUP_CF_TLS_API_ENV_VAR,
 };
 
-const CADDYFILE: &str = "Caddyfile";
-const PID_FILE: &str = "caddy-pid";
-const LOG_FILE: &str = "caddy-log";
+use super::{local_server::LINKUP_LOCAL_SERVER_PORT, BackgroundService};
 
-pub fn start(domains: Vec<String>) -> Result<()> {
-    if std::env::var(LINKUP_CF_TLS_API_ENV_VAR).is_err() {
-        return Err(CliError::StartCaddy(format!(
-            "{} env var is not set",
-            LINKUP_CF_TLS_API_ENV_VAR
-        )));
-    }
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Failed while handing file: {0}")]
+    FileHandling(#[from] std::io::Error),
+    #[error("Missing Cloudflare TLS API token on the environment variables")]
+    MissingTlsApiTokenEnv,
+    #[error("Redis shared storage is a new feature! You need to uninstall and reinstall local-dns to use it.")]
+    MissingRedisInstalation,
+}
 
-    let domains_and_subdomains: Vec<String> = domains
-        .iter()
-        .map(|domain| format!("{domain}, *.{domain}"))
-        .collect();
-
-    write_caddyfile(&domains_and_subdomains)?;
-
-    // Clear previous log file on startup
-    fs::write(linkup_file_path(LOG_FILE), "").map_err(|err| {
-        CliError::WriteFile(format!(
-            "Failed to clear log file at {}, error: {}",
-            linkup_file_path(LOG_FILE).display(),
-            err
-        ))
-    })?;
-
-    Command::new("caddy")
-        .current_dir(linkup_dir_path())
-        .arg("start")
-        .arg("--pidfile")
-        .arg(linkup_file_path(PID_FILE))
-        .stdout(Stdio::null())
-        .stderr(Stdio::null())
-        .status()
-        .map_err(|err| CliError::StartCaddy(err.to_string()))?;
-
-    Ok(())
+pub struct Caddy {
+    domains: Vec<String>,
+    caddyfile_path: PathBuf,
+    logfile_path: PathBuf,
+    pidfile_path: PathBuf,
 }
 
-pub fn stop() -> Result<()> {
-    Command::new("caddy")
-        .current_dir(linkup_dir_path())
-        .arg("stop")
-        .stdout(Stdio::null())
-        .stderr(Stdio::null())
-        .status()
-        .map_err(|err| CliError::StopErr(err.to_string()))?;
+impl Caddy {
+    pub fn new(domains: Vec<String>) -> Self {
+        Self {
+            domains,
+            caddyfile_path: linkup_file_path("Caddyfile"),
+            logfile_path: linkup_file_path("caddy-log"),
+            pidfile_path: linkup_file_path("caddy-pid"),
+        }
+    }
```

Comment:
**_note to self:_**
To make this easier to test, we can have another constructor with config to pass these paths params.